Addition of optional argument `on_verify_fail` to `check_all` and `check_error`

Karl-Michaud opened the pull request:
## Proposed Changes

As suggested by the PR title, I added the optional  argument `on_verify_fail` to `check_all` and `check_error` functions (and by extension the helper `_check`) in  `python_ta/check/__init__.py`.

The optional argument `on_verify_fail` can have:
- Default value "log" which keeps the default behaviour when file can't be checked (behaviour before this PR)
- "raise" which immediately raises a `ValueError` and stops execution when file can't be checked
- Else, raise `ValueError` which stops execution since `on_verify_fail` has invalid value

Moreover, since the task didn't specify what type of error should be raised, I decided that the easiest course of action would be to simply to raise a `ValueError`.


```python
if not verify_pre_check(file_py, linter.config.allow_pylint_comments):
    # File cannot be checked
    if on_verify_fail == "raise":
        # Flag to raise error has been chosen
        raise ValueError(f"File cannot be checked: {file_py}")
    # else statement not necessary, but reinforces that "log" keeps default behaviour
    elif on_verify_fail == "log":
        continue
    # Invalid flag
    else:
        raise ValueError(f"on_verify_fail cannot take value: {on_verify_fail}")
```

## Type of Change


| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |     X    |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |     X     |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist


Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] ~~I have updated the project documentation, if applicable.~~
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] ~~If this is my first contribution, I have added myself to the [list of contributors](https://github.com/pyta-uoft/pyta/blob/master/README.md#contributors).~~

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

I've updated some docstrings. Should I have also updated the HTML documentation?

Also, for the new test case, I was only able to find one file which triggered `verify_pre_check` to return False. It would be great to create more "bad" files in the future to be more thorough with the testing. 

Finally, it seems that `verify_pre_check` didn't return False for files that should have triggered it. Perhaps, It is an issue to be investigated in the future.

Any tips/criticism is highly appreciated and welcome. I'm always looking to improve!

david-yz-liu commented on python_ta/__init__.py:
```diff
@@ -63,6 +63,7 @@ def check_errors(
     output: Optional[Union[str, IO]] = None,
     load_default_config: bool = True,
     autoformat: Optional[bool] = False,
+    on_verify_fail: str = "log",
```
A more appropriate type annotation is `Literal`(https://docs.python.org/3/library/typing.html#typing.Literal)

david-yz-liu commented on python_ta/__init__.py:
```diff
@@ -106,6 +109,10 @@ def check_all(
         autoformat:
             If True, autoformat all modules using the black formatting tool before analyzing code.
+        on_verify_fail:
+            determines how to handle files that cannot be checked. If set to "log" (default), an error
```
Capitalize "Determines"

david-yz-liu commented on tests/test_check.py:
```diff
@@ -256,6 +265,50 @@ def test_check_no_reporter_output(
+            input_file,
+            config={
+                "output-format": "pyta-plain",
+                "pyta-error-permission": "no",
```
It's not necessary to add `pyta-error-permission`/`pyta-file-permission`, as the default values are `"no"`. I think other tests containing them are a holdover from previous behaviour.

david-yz-liu commented on python_ta/__init__.py:
```diff
@@ -155,7 +166,17 @@ def _check(
+                    # File cannot be checked
+                    if on_verify_fail == "raise":
+                        # Flag to raise error has been chosen
+                        raise ValueError(f"File cannot be checked: {file_py}")
```
Overall, this approach loses information that `verify_pre_check` already has access to. Look at the implementation of that function: it's raising errors, then catching them to return `False`. You can configure that function to either return `False` or re-raise the error using a _bare raise_.

Karl-Michaud replied:
Thanks for the input. I opted to modify `verify_pre_check` by adding the optional argument. Through that, I simply checked if an error should be raised (or not).

Karl-Michaud commented:
Added new changes. I opted to modify verify_pre_check by adding the optional argument. Through that, I simply checked if an error should be raised (or not).

This preserves all raised error data produced in `verify_pre_check`. Hence, in `_check`, if an error is raised, the function immediately jumps to the `except` branch and terminates when `on_verify_fail="raise"`. On the other hand, if  `on_verify_fail="log"` (default), this will preserve the default behaviour since False is returned.

Furthermore, since I modified `verify_pre_check`, I added some tests for better coverage in `tests/test_init_logging.py`.

Finally, due to the new use of `Literal["log", "raise"]`, I removed a useless test.
